packages/compass-aggregations/src/components/stage-editor/stage-editor.tsx

```diff
@@ -0,0 +1,239 @@
+  serverVersion: string;
+  autocompleteFields: CompletionWithServerInfo[];
+  syntaxError: PipelineParserError | null;
+  serverError: MongoServerError | null;
+  num_stages: number;
+  editor_view_type: string;
+  className: string;
+  onChange: (index: number, value: string) => void;
+};
+
+
+function useStageCompleter(
+  ...args: ConstructorParameters<typeof StageAutoCompleter>
+): StageAutoCompleter {
+  const [version, textCompleter, fields, operator] = args;
+  const completer = useRef<StageAutoCompleter>();
+  if (!completer.current) {
+    completer.current = new StageAutoCompleter(
+      version,
+      textCompleter,
+      fields,
+      operator
+    );
+  }
+  useEffect(() => {
+    completer.current?.update(fields, operator ?? null, version);
+  }, [fields]);
+  return completer.current;
+}
+
+const UnthemedStageEditor = ({
+  stageValue,
+  stageOperator,
+  index,
+  onChange,
+  serverError,
+  syntaxError,
+  darkMode,
+  className,
+  autocompleteFields,
+  serverVersion,
+  num_stages,
+  editor_view_type,
+}: StageEditorProps) => {
+  const editorInitialValueRef = useRef<string | null>(stageValue);
+  const editorRef = useRef<AceEditor | undefined>(undefined);
+  const completer = useStageCompleter(
+    serverVersion,
+    EditorTextCompleter,
+    autocompleteFields,
+    stageOperator
+  );
+
+  // When the index changes and the value is different from the editor value,
+  // update the editor value.
+  useEffect(() => {
+    if (editorRef.current?.getValue() !== stageValue) {
+      editorRef.current?.setValue(stageValue ?? '');
+    }
+  }, [index]);
+
+  useEffect(() => {
+    editorRef.current?.focus();
+  }, [stageOperator]);
+
+  useEffect(() => {
+    let annotations: AceAnnotation[] = [];
+    if (syntaxError && syntaxError.loc) {
+      const { line: row, column } = syntaxError.loc;
+      annotations = [
+        {
+          row: row - 1,
+          column,
+          text: syntaxError.message,
+          type: 'error',
+        },
+      ];
+    }
+    editorRef.current?.getSession().setAnnotations(annotations);
+  }, [syntaxError, editorRef]);
+
+  const onBlurEditor = useCallback(() => {
+    const value = editorRef.current?.getValue();
+    if (value !== undefined && value !== editorInitialValueRef.current) {
+      track('Aggregation Edited', {
+        num_stages: num_stages,
+        stage_index: index + 1,
+        stage_action: 'stage_content_changed',
+        stage_name: stageOperator,
+        editor_view_type: editor_view_type,
+      });
+      editorInitialValueRef.current = value;
+    }
+  }, [editorRef, editorInitialValueRef, num_stages, index, stageOperator, editor_view_type]);
+
+  return (
+    <div
+      className={cx(
+        editorContainerStyles,
+        darkMode
+          ? editorContainerStylesDark
+          : editorContainerStylesLight,
+        className
+      )}
+    >
+      <div className={editorStyles}>
+        <Editor
+          text={stageValue ?? undefined}
```

Comment:
Okay, so I took a closer look. The component is getting its props correctly. I think our root cause is here. Passing `null` or `undefined` here makes ace-editor just not update preserving the old value (uncontrolled behavior), this should be a fallback to the empty string instead:

```suggestion
          text={stageValue ?? ''}
```